src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -491,13 +492,26 @@ public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompl
             CancellationTokenSource cts;
             bool disposeCts;
             bool hasTimeout = _timeout != s_infiniteTimeout;
+            TimeoutState timeoutState = null; // Lazy initialization to avoid unnecessary alocation when timeout is not set
+            Timer timer = null;
             if (hasTimeout || cancellationToken.CanBeCanceled)
             {
                 disposeCts = true;
                 cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _pendingRequestsCts.Token);
                 if (hasTimeout)
                 {
-                    cts.CancelAfter(_timeout);
+                    timeoutState = new TimeoutState(cts);
```

Comment:
Rather than changing any of this, did you instead consider just starting accessing Environment.TickCount at the beginning of the request, and if cancellation is requested and Environment.TickCount - original time is > the timeout specified, treating it as a timeout?

Reply:
In that case, we can accidentally handle a caller's cancellation as a timeout which can lead to applying timeout-specific retry logic to a manual cancellation on the application level.
In example,
1) User click "Cancel" button on a page
2) Cancellation is propagated to some background process on the back-end (e.g. requesting data from another service)
3) Back-end code cancels the caller's token
4) Environment.TickCount - originalTime get larger than timeout
5) TaskCancelledException is thrown and caught by this logic
6) TimeoutException is thrown
7) Application code tries to retry the request due to timeout

Reply:
Actually, the above example is not precise since we can check the caller's token. The following would be better:
1. One of Socket's timeout triggered
2. Environment.TickCount - originalTime get larger than timeout
3. TaskCancelledException is thrown on the socket level and caught by this logic
4. TimeoutException is thrown
5. Application code tries to retry the request due to timeout
It's better than above, but now we have got a question. Is acceptable to treat Socket-level timeout as HttpClient one?
Secondly, the TickCount is Int32 and gets [reset each 24.9 days according to the docs.](https://docs.microsoft.com/en-us/dotnet/api/system.environment.tickcount?view=netcore-3.1)

Reply:
> Secondly, the TickCount is Int32 and gets reset each 24.9 days according to the docs.

If it's a concern, you can use Environment.TickCount64.

> we can accidentally handle a caller's cancellation as a timeout

If there are multiple reasons something is canceled, I'm not concerned about distinguishing them.  If an OCE is thrown, and we discover that we've exceeded the timeout, it's totally reasonable to me in that case to say "this timed out", even if other things caused it to be canceled as well.  Now, we might choose to prioritize certain sources if we can, e.g. we could treat it as non-timeout if we see that the user's token is already canceled, but at the end of the day race conditions here abound, and who's to say what the actual trigger was.

Reply:
I see your point, and mostly agree now since it's a pretty simple way to avoid allocating TimeoutState. However, I'm still concerned if we can solidly rely on Environment.TickCount64? I don't have any arguments here, just gut feeling.

Reply:
CancelAfter has been brought back and Timer was removed. Timing out is checked based on Environment.TickCount64 property.